proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+kindPolyAlphaTyVarSpec, kindPolyBetaTyVarSpec :: TyVarBinder
+kindPolyAlphaTyVarSpec = mkTyVarBinder Specified kindPolyAlphaTyVar
+kindPolyBetaTyVarSpec  = mkTyVarBinder Specified kindPolyBetaTyVar
+
+kindPolyAlphaTy, kindPolyBetaTy :: Type
+kindPolyAlphaTy = mkTyVarTy kindPolyAlphaTyVar
+kindPolyBetaTy  = mkTyVarTy kindPolyBetaTyVar
+```
+
+## Backward Compatibility
+
+0. No breakage should be expected for existing code.
+
+## Alternatives
+
+1. [More intuitive signature for `reifyInfoTablePtr#` primitive](#more-intuitive-signature-for-reifyinfotableptr-primitive)
+2. [Replace `LiftedCtorToSymbol` with a type family that directly returns the type-level list of fields of the constructor](#replace-liftedctortosymbol-with-a-type-family-that-directly-returns-the-type-level-list-of-fields-of-the-constructor)
+3. [Remove `LiftedCtorToSymbol` built-in typeclass entirely](#remove-liftedctortosymbol-built-in-typeclass-entirely)
+4. [Integrate the library code for safe DPS structure building into GHC](#integrate-the-library-code-for-safe-dps-structure-building-into-ghc)
+5. [Add nothing to GHC, and relie on unsafe user-side tricks to achieve roughly the same behavior](#add-nothing-to-ghc-and-relie-on-unsafe-user-side-tricks-to-achieve-roughly-the-same-behavior)
+
+### More intuitive signature for `reifyInfoTablePtr#` primitive
+
+`reifyInfoTablePtr#` has a rather counterintuitive signature:
+
+```haskell
+reifyInfoTablePtr# :: forall k (a :: k). (# #) -> InfoTablePtrOf# a
+```
+
+This signature is motivated by the fact that in `emitPrimOp`, we only have access to the type of the return value of the primop, and not to the types of the arguments. So in the current design, I feed the type-level information of the chosen constructor as a (phantom) type argument on the return type of the primop.
+
+It would be clearer to have signature like:
+
+```haskell
+reifyInfoTablePtrAlt# :: forall k (a :: k). Proxy# a -> Addr#
+```
+
+but that would require more important restructuration of the primop code generation pipeline, and I don't think it's worth it.
+
+### Replace `LiftedCtorToSymbol` with a type family that directly returns the type-level list of fields of the constructor
+
+`LiftedCtorToSymbol` is only used as a way to extract information about a constructor's fields from the `GHC.Generics` representation of the associated type. We could devise a typeclass that does all of that directly into the compiler, without requiring `GHC.Generics`.
+
+I expect that it would require to sort of reimplement part of `GHC.Generics` somewhere else in the compiler. I haven't looked into that yet.
+
+### Remove `LiftedCtorToSymbol` built-in typeclass entirely
+
+With the proposed changes, the library code can have syntax like
+
+```haskell
+fill @'MkIdCard d
+```
+
+and still recover information about `MkIdCard` from the `GHC.Generics` representation of `IdCard`.
+
+Without the new built-in typeclass, the syntax would become:
+
+```haskell
+fill @'MkIdCard @"MkIdCard" d
```

Comment:
Could it be `fill @IdCard @"MkIdCard"` instead? Still not great, but at least not repetitive.

Reply:
If `fill` no longer receive the constructor lifted as a type (`'Left`), then we need to implement the resolution `(name :: Symbol, baseType :: Type) -> DataCon` inside `reifyInfoTablePtr#`. That means a more complex `reifyInfoTablePtr#`, with little advantage (still 2 type parameters to supply)